tgui/packages/tgui/components/NumberInput.tsx

```diff
@@ -0,0 +1,351 @@
+      return;
+    }
+
+    this.setState((prevState) => {
+      const state = { ...prevState };
+      const offset = state.origin - event.screenY;
+      if (prevState.dragging && step) {
+        const stepOffset = Number.isFinite(minValue) ? minValue % step : 0;
+        // Translate mouse movement to value
+        // Give it some headroom (by increasing clamp range by 1 step)
+        state.internalValue = clamp(
+          state.internalValue + (offset * step) / (stepPixelSize || 1),
+          minValue - step,
+          maxValue + step,
+        );
+        // Clamp the final value
+        state.value = clamp(
+          state.internalValue - (state.internalValue % step) + stepOffset,
+          minValue,
+          maxValue,
+        );
+        state.origin = event.screenY;
+      } else if (Math.abs(offset) > 4) {
+        state.dragging = true;
+      }
+      return state;
+    });
+  };
+
+  handleDragEnd: MouseEventHandler<HTMLInputElement> = (event) => {
+    const { value, dragging, internalValue } = this.state;
+    const { onDrag, onChange, disabled } = this.props;
+    if (disabled) {
+      return;
+    }
+
+    clearInterval(this.dragInterval);
+    clearTimeout(this.dragTimeout);
+    this.setState({
+      dragging: false,
+      editing: !dragging,
+    });
+
+    if (dragging) {
+      this.suppressFlicker();
+      if (onChange) {
+        onChange(event, value);
+      }
+      if (onDrag) {
+        onDrag(event, value);
+      }
+    } else if (this.inputRef) {
+      const input = this.inputRef.current;
+      if (input) {
+        input.value = `${internalValue}`;
+        // IE8: Dies when trying to focus a hidden element
+        // (Error: Object does not support this action)
+        setTimeout(() => {
+          try {
+            input.focus();
+            input.select();
+          } catch {}
+        }, 1);
+      }
+    }
+  };
+
+  handleBlur: FocusEventHandler<HTMLInputElement> = (event) => {
+    const { editing } = this.state;
+    const { minValue, maxValue, onChange, onDrag, disabled } = this.props;
+
+    if (disabled || !editing) {
+      return;
+    }
+
+    const targetValue = clamp(
+      parseFloat(event.target.value),
+      minValue,
+      maxValue,
+    );
+    if (Number.isNaN(targetValue)) {
+      this.setState({
+        editing: false,
+      });
+      return;
+    }
+    this.setState({
+      editing: false,
+    });
+    this.suppressFlicker();
+    if (onChange) {
+      onChange(event, targetValue);
+    }
+    if (onDrag) {
+      onDrag(event, targetValue);
+    }
+  };
+
+  handleKeyDown: KeyboardEventHandler<HTMLInputElement> = (event) => {
+    const { minValue, maxValue, onChange, onDrag, disabled } = this.props;
+    if (disabled) {
+      return;
+    }
+
+    if (event.key === KEY.Enter) {
+      const targetValue = clamp(
+        parseFloat(event.currentTarget.value),
+        minValue,
+        maxValue,
+      );
+      if (Number.isNaN(targetValue)) {
```

Comment:
i think isNaN is part of the standard library - do we need the `Number.` part? correct me if I'm wrong here pls

Reply:
Yup Number. is not required 👍